### ISSUE TITLE
Does MGTwoLevelTransfer work for any other vector than LinearAlgebra::distributed::Vector<Number>?

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#16684]
I am a bit confused by the `MGTwoLevelTransfer` class. Its specialization for `LinearAlgebra::distributed::Vector<Number>` makes sense to me, but the general template is declared through an abstract base class
```
template <typename VectorType>
class MGTwoLevelTransferBase : public Subscriptor
{
public:
  /**
   * Perform prolongation on a solution vector.
   */
  virtual void
  prolongate_and_add(VectorType &dst, const VectorType &src) const = 0;

  /**
   * Perform restriction on a residual vector.
   */
  virtual void
  restrict_and_add(VectorType &dst, const VectorType &src) const = 0;

  /**
   * Perform interpolation of a solution vector from the fine level to the
   * coarse level. This function is different from restriction, where a
   * weighted residual is transferred to a coarser level (transposition of
   * prolongation matrix). In other words, restriction acts on right hand
   * side vectors, whereas interpolation acts on solution vectors.
   */
  virtual void
  interpolate(VectorType &dst, const VectorType &src) const = 0;

  /**
   * Enable inplace vector operations if external and internal vectors
   * are compatible.
   */
  virtual void
  enable_inplace_operations_if_possible(
    const std::shared_ptr<const Utilities::MPI::Partitioner>
      &partitioner_coarse,
    const std::shared_ptr<const Utilities::MPI::Partitioner>
      &partitioner_fine) = 0;

  /**
   * Return the memory consumption of the allocated memory in this class.
   */
  virtual std::size_t
  memory_consumption() const = 0;
};
```
and a derived class:
```
template <int dim, typename VectorType>
class MGTwoLevelTransfer : public MGTwoLevelTransferBase<VectorType>
{
public:
  /**
   * @copydoc MGTwoLevelTransferBase::prolongate_and_add
   */
  void
  prolongate_and_add(VectorType &dst, const VectorType &src) const override;

  /**
   * @copydoc MGTwoLevelTransferBase::restrict_and_add
   */
  void
  restrict_and_add(VectorType &dst, const VectorType &src) const override;

  /**
   * @copydoc MGTwoLevelTransferBase::interpolate
   */
  void
  interpolate(VectorType &dst, const VectorType &src) const override;

  /**
   * Enable inplace vector operations if external and internal vectors
   * are compatible.
   */
  void
  enable_inplace_operations_if_possible(
    const std::shared_ptr<const Utilities::MPI::Partitioner>
      &partitioner_coarse,
    const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner_fine)
    override;

  /**
   * Return the memory consumption of the allocated memory in this class.
   */
  std::size_t
  memory_consumption() const override;
};
```

But I do not think that these functions are actually implemented anywhere -- and even if they were, I wouldn't understand *how* because neither the base nor the derived class ever learns about about the involved `DoFHandler` objects through constructors or function arguments.

In other words, are these two classes basically just dead declarations?

**Comment by kronbichler** [comment]
I don't think those other types would work (or have been envisioned at this point), so the declarations are dead indeed - they originate from the other `MGTransfer` classes in the (old) local smoothing infrastructure, where we have support for both a matrix-based choice (the setting of the base class listed above) and the specialization for `LinearAlgera::distributed::Vector` that stores the indices in the MPI-local index space and then can perform direct array access to read/write entries from vectors using `LinearAlgebra::distributed::Vector::local_element` or the respective pointers. I have no solution to this, apart from the desire that we should try to unify the multigrid infrastructure to a single instance for both global coarsening and local smoothing. @peterrum has been working on it primarily, and I keep helping a little here and there. I think it will take some time to finalize, and this issue can remind us of it.

**Comment by bangerth** [comment]
Should we just disable the existing general template for now, to make it clear that it is not usable?

**Comment by kronbichler** [comment]
I think that would be a good option. @peterrum opinions?

**Comment by peterrum** [comment]
I am fine with removing these.
